// zca-js/dist/cjs/apis/blockUser.cjs
'use strict';

var ZaloApiError = require('../Errors/ZaloApiError.cjs');
var context = require('../context.cjs');
var utils = require('../utils.cjs');

function blockUserFactory(serviceURL) {
    return async function blockUser(userId) {
        if (!context.appContext.secretKey)
            throw new ZaloApiError.ZaloApiError("Secret key is not available");
        if (!context.appContext.imei)
            throw new ZaloApiError.ZaloApiError("IMEI is not available");
        if (!context.appContext.cookie)
            throw new ZaloApiError.ZaloApiError("Cookie is not available");
        if (!context.appContext.userAgent)
            throw new ZaloApiError.ZaloApiError("User agent is not available");
        if (!context.appContext.language)
            throw new ZaloApiError.ZaloApiError("Language is not available");
        const params = {
            fid: userId,
            imei: context.appContext.imei,
        };
        const encryptedParams = utils.encodeAES(context.appContext.secretKey, JSON.stringify(params));
        if (!encryptedParams)
            throw new ZaloApiError.ZaloApiError("Failed to encrypt params");
        const response = await utils.request(serviceURL, {
            method: "POST",
            body: new URLSearchParams({
                params: encryptedParams,
            }),
        });
        const result = await utils.handleZaloResponse(response);
        if (result.error)
            throw new ZaloApiError.ZaloApiError(result.error.message, result.error.code);
        return result.data;
    };
}

exports.blockUserFactory = blockUserFactory;
